### PR TITLE
[FLINK-13930][hive] Support Hive version 3.1.x

### PR DESCRIPTION
**Pull request by xuefuz**
## What is the purpose of the change

Support Hive 3.1.x versions, including 3.1.0, 3.1.1, and 3.1.2. Those are latest and newest versions from Hive community.

## Brief change log

  - Added new shims for each version
  - Made inheritance relationships between the shims
  - Shimmed methods to deal with API difference

## Verifying this change

This change is already covered by existing tests, and manually ran the tests for all versions.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by xuefuz** [comment]
cc: @bowenli86 @lirui-apache @zjuwangg

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSimpleUDF.java [hunk 1]
```diff
@@ -127,9 +130,10 @@ public DataType getHiveResultType(Object[] constantArguments, DataType[] argType
 				.getResolver().getEvalMethod(argTypeInfo).getReturnType();
 
 			return HiveInspectors.toFlinkType(
-				HiveInspectors.getObjectInspector(returnType));
+				HiveInspectors.getObjectInspector(hiveShim, returnType));
 		} catch (UDFArgumentException e) {
 			throw new FlinkHiveUDFException(e);
 		}
 	}
+
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSimpleUDF.java]: revert？

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java [hunk 2]
```diff
@@ -121,30 +124,30 @@ public void testUDFJson() {
 		assertEquals(null, udf.eval(json, pattern));
 	}
 
-	@Test
-	public void testUDFMinute() {
+//	@Test Disabled because in HIve 3.1.0 UDFMinute is no long a simple UDF.
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java]: I'm fine just removing this test or replace it with another Hive UDF, instead of disabling it

Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java]: Ok. I will remove it.

**Comment by xuefuz** [comment]
@bowenli86 Please take another look when you get a chance.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDF.java [hunk 3]
```diff
@@ -40,10 +41,11 @@
 	private static final Logger LOG = LoggerFactory.getLogger(HiveGenericUDF.class);
 
 	private transient GenericUDF.DeferredObject[] deferredObjects;
+	private transient HiveShim hiveShim;
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDF.java]: Does this mean hiveShim is not needed after deserialization?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java [hunk 4]
```diff
@@ -122,29 +124,16 @@ public void testUDFJson() {
 	}
 
 	@Test
-	public void testUDFMinute() {
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java]: why is this removed?

Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/functions/hive/HiveSimpleUDFTest.java]: UDFMinute isn't a class UDF any more, so the test is invalid in 3.1.x.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShim.java [hunk 5]
```diff
@@ -109,4 +112,53 @@ void alterPartition(IMetaStoreClient client, String databaseName, String tableNa
...
+	 * @throws IOException in case of any io error
+	 */
+	FileStatus[] getFileStatusRecurse(Path path, int level, FileSystem fs) throws IOException;
+
+	/**
+	 * The signature of HiveStatsUtils.makeSpecFromName() was changed in Hive 3.1.0.
+	 *
+	 * @param partSpec partition specs
+	 * @param currPath the current path
+	 * @param notUsed newly added argument (which we don't care)
```
[the remark this answers is not in the document]
Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShim.java]: good point.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDF.java (hunk 3 above)
Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDF.java]: shim is only needed at initialization at the planner. also, shim is not serializable.

Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDF.java]: I created https://issues.apache.org/jira/browse/FLINK-13947 to investigate this.

**Comment by xuefuz** [comment]
PR updated based on review feedback. @lirui-apache @bowenli86 Could you please take another look? Thanks.

**Comment by bowenli86** [comment]
The python test failed in CI, but master has been green. Can you double check what's going on with python?

**Comment by xuefuz** [comment]
Rebasing to latest master solved the python build issue, @bowenli86

**Comment by bowenli86** [comment]
LGTM, merging
